lib/src/Pos/Generator/Block/Logic.hs

```diff
@@ -60,22 +59,32 @@ type BlockTxpGenMode g ctx m =
 -- | Generate an arbitrary sequence of valid blocks. The blocks are
 -- valid with respect to the global state right before this function
 -- call.
+-- The blocks themselves can be combined and retained according to some monoid.
+-- Intermediate results will be forced. Blocks can be generated, written to
+-- disk, then collected by using '()' as the monoid and 'const ()' as the
+-- injector, for example.
 genBlocks ::
-       forall g ctx m . BlockTxpGenMode g ctx m
+       forall g ctx m t . (BlockTxpGenMode g ctx m, Monoid t)
     => BlockGenParams
-    -> RandT g m (OldestFirst [] Blund)
-genBlocks params = do
+    -> (Maybe Blund -> t)
+    -> RandT g m t
+genBlocks params inj = do
     ctx <- lift $ mkBlockGenContext @(MempoolExt m) params
     mapRandT (`runReaderT` ctx) genBlocksDo
   where
-    genBlocksDo =
-        OldestFirst <$> do
-            let numberOfBlocks = params ^. bgpBlockCount
-            tipEOS <- getEpochOrSlot <$> lift getTipHeader
-            let startEOS = succ tipEOS
-            let finishEOS =
-                    toEnum $ fromEnum tipEOS + fromIntegral numberOfBlocks
-            catMaybes <$> mapM genBlock [startEOS .. finishEOS]
+    genBlocksDo = do
+        let numberOfBlocks = params ^. bgpBlockCount
+        tipEOS <- getEpochOrSlot <$> lift getTipHeader
+        let startEOS = succ tipEOS
+        let finishEOS = toEnum $ fromEnum tipEOS + fromIntegral numberOfBlocks
+        foldM' genOneBlock mempty [startEOS .. finishEOS]
+
+    genOneBlock t eos = ((t <>) . inj) <$> genBlock eos
+
+    foldM' _ !base [] = return base
+    foldM' combine !base (x : xs) = do
```

Comment:
What about applying a worker-wrapper transformation here, so that `combine` is not passed at each step? Could be good for both perfomance and clarity.

```
foldM' combine = go
  where
    go !base [] = return base
    go !base (x:xs) = do
      it <- combine base x
      go it xs
```